python/ml4ir/applications/ranking/model/layers/masking.py

```diff
@@ -159,6 +166,105 @@ def __init__(self,
+                           dtype=tf.float32)
+            mask = tf.expand_dims(mask, axis=1)
+            mask = tf.tile(mask, [1, sequence_len, 1])
+            return tf.multiply(inputs, mask)
+        else:
+            return inputs
+
+    def apply_fixed_mask(self, inputs, training):
+        """
+        Apply a fixed mask to the input tensor.
+
+        This function applies a fixed mask to the input tensor based on pre-generated
+        masks. If the masks are not yet created, it initializes them. The mask is applied
+        by element-wise multiplication with the input tensor.
+
+        Parameters:
+        inputs (tf.Tensor): A tensor of shape (batch_size, sequence_len, num_features)
+                            representing the input data.
+        training (bool): A boolean indicating whether the model is in training mode.
+
+        Returns:
+        tf.Tensor: A tensor of the same shape as `inputs` with the fixed mask applied.
+
+        Notes:
+        - If the fixed masks are not initialized, they are created using `create_fixed_masks`.
+        - The function updates `current_mask_index` to cycle through the masks for each call.
+
+        Example:
+        >>> inputs = tf.random.uniform((2, 3, 4))
+        >>> masked_inputs = apply_fixed_mask(inputs, training=True)
+        >>> masked_inputs.shape
+        TensorShape([2, 3, 4])
+        """
+        batch_size, sequence_len, num_features = tf.shape(inputs)[0], tf.shape(inputs)[1], tf.shape(inputs)[2]
+        if self.fixed_masks is None:
+            self.fixed_masks = self.create_fixed_masks(num_features)
+            QueryFeatureMask.masking_config[MonteCarloInferenceKey.FIXED_MASK_COUNT] = len(self.fixed_masks)
+
+        current_mask = tf.expand_dims(tf.expand_dims(self.fixed_masks[self.current_mask_index], 0), 0)
+        current_mask = tf.tile(current_mask, [batch_size, sequence_len, 1])
+        self.current_mask_index += 1
+        self.current_mask_index %= QueryFeatureMask.masking_config[MonteCarloInferenceKey.FIXED_MASK_COUNT]
+        return tf.multiply(inputs, tf.cast(current_mask, tf.float32))
+
+    def create_fixed_masks(self, num_features):
+        """
+        Generate all possible fixed masks for a given feature dimension.
+
+        This function creates a list of tuples representing all possible combinations
+        of 0s and 1s for a specified feature dimension. Each tuple corresponds to a
+        fixed mask.
+
+        Parameters:
+        num_features (int): The dimensionality of the feature space, representing
+                           the length of each tuple.
+
+        Returns:
+        list of tuples: A list containing all possible fixed masks, where each mask
+                        is a tuple of length `num_features` with elements 0 or 1.
+
+        Example:
+        >>> create_fixed_masks(2)
+        [(0, 0), (0, 1), (1, 0), (1, 1)]
+        """
+        return list(itertools.product([0, 1], repeat=num_features))
```

Comment:
@mohazahran We should probably leave a todo to accept an arg from user config on how many features you want to pick in each mask? For example, if there are 4 features and we want to use 2 each time, you'd then get 4C2 masks. The code here is fine, it might blow up for longer feature sets.

For example, for 10 features you have 1024 masks. We should also add a flag for selecting n masks at random.

Anyway, this is for the future.

NOTE - this function also creates a mask with all 0s - (0, 0, 0, 0)

Reply:
In my mind, 4 features and we want to use 2 each time. Then use only the 2 features to get full masks (2^2).

That's true it will blow up by increasing the number of features and we would need to sample the full combinations and fix those.

Reply:
- NOTE - this function also creates a mask with all 0s - (0, 0, 0, 0)
That's true, this part I was in split. I can remove this mask from the mask set. However, since we are possibly just using a subset of all the features in masking, having all zero mask would force the model to only use features not in the mask. which might be beneficial in certain situations.